src/System/Random/Internal.hs

```diff
@@ -0,0 +1,1040 @@
+  uniformRM (CUIntMax b, CUIntMax t) = fmap CUIntMax . uniformRM (b, t)
+  {-# INLINE uniformRM #-}
+
+-- | See [Floating point number caveats](System-Random-Stateful.html#fpcaveats).
+instance UniformRange CFloat where
+  uniformRM (CFloat l, CFloat h) = fmap CFloat . uniformRM (l, h)
+  {-# INLINE uniformRM #-}
+
+-- | See [Floating point number caveats](System-Random-Stateful.html#fpcaveats).
+instance UniformRange CDouble where
+  uniformRM (CDouble l, CDouble h) = fmap CDouble . uniformRM (l, h)
+  {-# INLINE uniformRM #-}
+
+
+-- The `chr#` and `ord#` are the prim functions that will be called, regardless of which
+-- way you gonna do the `Char` conversion, so it is better to call them directly and
+-- bypass all the hoops. Also because `intToChar` and `charToInt` are internal functions
+-- and are called on valid character ranges it is impossible to generate an invalid
+-- `Char`, therefore it is totally fine to omit all the unnecessary checks involved in
+-- other paths of conversion.
+word32ToChar :: Word32 -> Char
+word32ToChar (W32# w#) = C# (chr# (word2Int# w#))
+{-# INLINE word32ToChar #-}
+
+charToWord32 :: Char -> Word32
+charToWord32 (C# c#) = W32# (int2Word# (ord# c#))
+{-# INLINE charToWord32 #-}
+
+instance Uniform Char where
+  uniformM g = word32ToChar <$> unbiasedWordMult32 (charToWord32 maxBound) g
+  {-# INLINE uniformM #-}
+instance UniformRange Char where
+  uniformRM (l, h) g =
+    word32ToChar <$> unbiasedWordMult32RM (charToWord32 l, charToWord32 h) g
+  {-# INLINE uniformRM #-}
+
+instance Uniform Bool where
+  uniformM = fmap wordToBool . uniformWord8
+    where wordToBool w = (w .&. 1) /= 0
+instance UniformRange Bool where
+  uniformRM (False, False) _g = return False
+  uniformRM (True, True)   _g = return True
+  uniformRM _               g = uniformM g
+
+-- | See [Floating point number caveats](System-Random-Stateful.html#fpcaveats).
+instance UniformRange Double where
+  uniformRM (l, h) g = do
+    w64 <- uniformWord64 g
+    let x = word64ToDoubleInUnitInterval w64
+    return $ (h - l) * x + l
+
+-- | Turns a given uniformly distributed 'Word64' value into a uniformly
+-- distributed 'Double' value in the range [0, 1].
+word64ToDoubleInUnitInterval :: Word64 -> Double
+word64ToDoubleInUnitInterval w64 = d / m
+  where
+    d = fromIntegral w64 :: Double
+    m = fromIntegral (maxBound :: Word64) :: Double
+{-# INLINE word64ToDoubleInUnitInterval #-}
+
+-- | See [Floating point number caveats](System-Random-Stateful.html#fpcaveats).
+instance UniformRange Float where
+  uniformRM (l, h) g = do
+    w32 <- uniformWord32 g
+    let x = word32ToFloatInUnitInterval w32
+    return $ (h - l) * x + l
+
+-- | Turns a given uniformly distributed 'Word32' value into a uniformly
+-- distributed 'Float' value in the range [0,1].
+word32ToFloatInUnitInterval :: Word32 -> Float
+word32ToFloatInUnitInterval w32 = f / m
+  where
+    f = fromIntegral w32 :: Float
+    m = fromIntegral (maxBound :: Word32) :: Float
```

Comment:
It does not matter whether you divide by `2^32-1` or by `2^32` (basically because `(2**32-1) / 2**32 :: Float == 1.0`), and same applies to `Double`. So potentially one can try to shave off divisions:

```haskell
foreign import ccall safe "__word_encodeFloat"
  word_encodeFloat :: Word32 -> Int -> Float

word32ToFloatInUnitInterval :: Word32 -> Float
word32ToFloatInUnitInterval = flip word_encodeFloat (-32)

foreign import ccall safe "__word_encodeDouble"
  word_encodeDouble :: Word64 -> Int -> Double

word64ToDoubleInUnitInterval :: Word64 -> Double
word64ToDoubleInUnitInterval = flip word_encodeDouble (-64)
```

`__word_encodeFloat` and `__word_encodeDouble` are provided by RTS.

Reply:
You are aware of all the discussions on floating point? E.g. https://github.com/idontgetoutmuch/random/pull/118. Can you post your performance and coverage figures?

Reply:
Doesn't look similar to me, I'm not proposing an `unsafeCoercish` approach.

Coverage remains exactly the same; the statement "It does not matter whether you divide by 2^32-1 or by 2^32" is not an exagerration. It is literally that `fromIntegral w / (2**32-1) == fromIntegral w / (2**32 :: Float)` for all `w :: Word32`.

Reply:
I'm not sure whether this approach is faster (benchmarking it at the moment). But FFI must be marked as unsafe. I got 10x speedup for erf by replacing safe with unsafe. `__word_encodeDouble` should be in the same ns range

Reply:
You are right it is different.

```
bash-3.2$ git checkout v1.2-proposal
Switched to branch 'v1.2-proposal'
Your branch is up to date with 'idontgetoutmuch/v1.2-proposal'.
bash-3.2$ stack bench random:bench --ba '--small --match prefix pure/random/'

Running 1 benchmarks...
Benchmark bench: RUNNING...
pure/random/Float                        mean 26.58 μs  ( +- 1.664 μs  )
pure/random/Double                       mean 26.54 μs  ( +- 1.080 μs  )
pure/random/Integer                      mean 314.5 μs  ( +- 9.997 μs  )

bash-3.2$ git checkout float-via-rts
Switched to branch 'float-via-rts'
bash-3.2$ stack bench random:bench --ba '--small --match prefix pure/random/'

Running 1 benchmarks...
Benchmark bench: RUNNING...
pure/random/Float                        mean 3.415 ms  ( +- 136.0 μs  )
pure/random/Double                       mean 3.290 ms  ( +- 74.07 μs  )
pure/random/Integer                      mean 313.7 μs  ( +- 7.777 μs  )

bash-3.2$ git diff v1.2-proposal float-via-rts

diff --git a/src/System/Random/Internal.hs b/src/System/Random/Internal.hs
index 556e86c..a41545b 100644
--- a/src/System/Random/Internal.hs
+++ b/src/System/Random/Internal.hs
@@ -716,12 +716,15 @@ instance UniformRange Double where
 -- | Turns a given uniformly distributed 'Word64' value into a uniformly
 -- distributed 'Double' value in the range [0, 1].
 word64ToDoubleInUnitInterval :: Word64 -> Double
-word64ToDoubleInUnitInterval w64 = d / m
-  where
-    d = fromIntegral w64 :: Double
-    m = fromIntegral (maxBound :: Word64) :: Double
+word64ToDoubleInUnitInterval = flip word_encodeDouble (-64)
 {-# INLINE word64ToDoubleInUnitInterval #-}
 
+foreign import ccall safe "__word_encodeFloat"
+  word_encodeFloat :: Word32 -> Int -> Float
+
+foreign import ccall safe "__word_encodeDouble"
+  word_encodeDouble :: Word64 -> Int -> Double
+
 -- | See /Floating point number caveats/ in "System.Random.Stateful".
 instance UniformRange Float where
   uniformRM (l, h) g = do
@@ -732,10 +735,7 @@ instance UniformRange Float where
 -- | Turns a given uniformly distributed 'Word32' value into a uniformly
 -- distributed 'Float' value in the range [0,1].
 word32ToFloatInUnitInterval :: Word32 -> Float
-word32ToFloatInUnitInterval w32 = f / m
-  where
-    f = fromIntegral w32 :: Float
-    m = fromIntegral (maxBound :: Word32) :: Float
+word32ToFloatInUnitInterval = flip word_encodeFloat (-32)
 {-# INLINE word32ToFloatInUnitInterval #-}
 
 -- The two integer functions below take an [inclusive,inclusive] range.
```

Reply:
Did you try it with unsafe FFI?

Reply:
> I'm not sure whether this approach is faster (benchmarking it at the moment). But FFI must be marked as unsafe. I got 10x speedup for erf by replacing safe with unsafe. `__word_encodeDouble` should be in the same ns range

I tried that but it's still about x20 slower

```
Running 1 benchmarks...
Benchmark bench: RUNNING...
pure/random/Float                        mean 517.0 μs  ( +- 19.29 μs  )
pure/random/Double                       mean 429.3 μs  ( +- 11.53 μs  )
pure/random/Integer                      mean 341.0 μs  ( +- 43.84 μs  )
```

Reply:
```
foreign import ccall unsafe "__word_encodeFloat"
  word_encodeFloat :: Word32 -> Int -> Float

foreign import ccall unsafe "__word_encodeDouble"
  word_encodeDouble :: Word64 -> Int -> Double
```

Reply:
Thanks for benchmarking. 
Interesting. I also tried `scaleFloat` and it is even worse. A bit embarassing that there is no fast way just to change magnitude bits.

Reply:
Ok if I mark this as resolved?